README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# debounce_textfield

A simple, lightweight, customizable textfield with debounce for preventing continuous execution of (onChange) callback.

## Installation

1. Add the latest version of package to your pubspec.yaml (and run`dart pub get`):

```yaml
dependencies:
  debounce_textfield: ^0.0.1
```

2. Import the package and use it in your Flutter App.

```dart
import 'package:debounce_textfield/debounce_textfield.dart';
```

## Image

![alt text](screenshot.jpg)

## Example

You can simply just use the `DebounceTextfield` class and provide a callback to be executed after the delay.

```dart
DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
        ),
```

## customization

### Custom Duration

```dart
DebounceTextfield(
          action: (enteredText) {
            print('Executed after 800 milliseconds from last change.');
            print(enteredText);
          },
          duration: const Duration(milliseconds: 800),
        ),
```

### onTextfieldEmpty Callback

```dart
 DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
          onTextfieldEmpty: () {
            print('Textfield is empty now...');
          },
        ),
```

### Custom Style

```dart
 DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
          height: 50,
          clearButtonIcon:
              const Icon(Icons.clear_rounded, color: Colors.white, size: 20),
          textFieldDecoration: BoxDecoration(
            borderRadius: BorderRadius.circular(16),
            color: Colors.indigo.shade200,
          ),
          inputDecoration: const InputDecoration(
            border: InputBorder.none,
            hintText: 'Search for your favoutie Anime',
            hintStyle: TextStyle(
              fontSize: 15,
              color: Colors.white,
            ),
            icon: Icon(
              Icons.search,
              color: Colors.white,
              size: 20,
            ),
          ),
          textFieldStyle: const TextStyle(
            fontSize: 15,
            color: Colors.white,
          ),
        ),
```

### Right-to-left Direction

```dart
 DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
          direction: TextDirection.rtl,
          inputDecoration: const InputDecoration.collapsed(
            hintText: 'أدخل النص هنا',
            hintStyle: TextStyle(
              fontSize: 13,
              color: Colors.black54,
            ),
          ),
          textFieldDecoration: BoxDecoration(
            borderRadius: BorderRadius.circular(16),
            color: Colors.black12,
          ),
        ),
```


CHANGELOG.md
## 0.0.1

initial release of the `debounce_textfield` package.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:debounce_textfield_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"debounce_textfield_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/src/debounce_textfield.dart
import 'package:flutter/material.dart';
import 'dart:async';

class DebounceTextfield extends StatefulWidget {
  Function action;
  VoidCallback? onTextfieldEmpty;
  Duration duration;
  double height;
  Widget leadingWidget;
  TextStyle textFieldStyle;
  BoxDecoration textFieldDecoration;
  InputDecoration inputDecoration;
  TextDirection direction;
  Widget clearButtonIcon;
  EdgeInsetsGeometry margin;
  EdgeInsetsGeometry padding;

  DebounceTextfield({
    required this.action,
    this.onTextfieldEmpty,
    this.duration = const Duration(milliseconds: 500),
    this.height = 40,
    this.leadingWidget = const SizedBox.shrink(),
    this.textFieldStyle = const TextStyle(fontSize: 13),
    this.textFieldDecoration = const BoxDecoration(color: Color(0xffE8E8E8)),
    this.inputDecoration = const InputDecoration.collapsed(
      hintText: 'Enter Your Text Here',
      hintStyle: TextStyle(fontSize: 13, color: Color(0xffaeaeae)),
    ),
    this.clearButtonIcon =
        const Icon(Icons.clear_rounded, color: Color(0xffaeaeae), size: 20),
    this.direction = TextDirection.ltr,
    this.margin = const EdgeInsets.fromLTRB(16, 8, 16, 8),
    this.padding = const EdgeInsets.fromLTRB(16, 8, 16, 8),
    Key? key,
  }) : super(key: key);

  @override
  State<DebounceTextfield> createState() => _DebounceTextfieldState();
}

class _DebounceTextfieldState extends State<DebounceTextfield> {
  late Function _action;
  late Duration _duration;
  late StreamController<bool> _btnClearController;
  late TextEditingController _inputController;
  Timer? _timer;

  @override
  void initState() {
    super.initState();
    _action = widget.action;
    _duration = widget.duration;
    _btnClearController = StreamController<bool>();
    _inputController = TextEditingController();
  }

  @override
  void dispose() {
    _btnClearController.close();
    _inputController.dispose();
    _timer?.cancel();
    super.dispose();
  }

  runDebouncer(String text) {
    _timer?.cancel();
    if (text.isNotEmpty) {
      _btnClearController.add(true);
      _timer = Timer(_duration, () => _action(text));
    } else {
      _btnClearController.add(false);
      if (widget.onTextfieldEmpty != null) {
        widget.onTextfieldEmpty!();
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: widget.direction,
      child: Container(
        height: widget.height,
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            widget.leadingWidget,
            const SizedBox(width: 5),
            Expanded(
              child: TextFormField(
                controller: _inputController,
                onChanged: (value) => runDebouncer(value),
                decoration: widget.inputDecoration,
                style: widget.textFieldStyle,
              ),
            ),
            StreamBuilder<bool>(
                stream: _btnClearController.stream,
                builder: (context, snapshot) {
                  return snapshot.data == true
                      ? GestureDetector(
                          onTap: () {
                            _inputController.clear();
                            _btnClearController.add(false);
                            if (widget.onTextfieldEmpty != null) {
                              widget.onTextfieldEmpty!();
                            }
                          },
                          child: widget.clearButtonIcon)
                      : const SizedBox.shrink();
                }),
          ],
        ),
        decoration: widget.textFieldDecoration,
        padding: widget.padding,
        margin: widget.margin,
      ),
    );
  }
}


example/lib/main.dart
import 'package:debounce_textfield/debounce_textfield.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Debounce Textfield Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Widget> example1() => [
        const Text('Default'),
        DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
        ),
      ];

  List<Widget> example2() => [
        const Text('Custom Duration'),
        DebounceTextfield(
          action: (enteredText) {
            print('Executed after 800 milliseconds from last change.');
            print(enteredText);
          },
          duration: const Duration(milliseconds: 800),
        ),
      ];

  List<Widget> example3() => [
        const Text('onTextfieldEmpty Callback'),
        DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
          onTextfieldEmpty: () {
            print('Textfield is empty now...');
          },
        ),
      ];

  List<Widget> example4() => [
        const Text('Custom Style'),
        DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
          height: 50,
          clearButtonIcon:
              const Icon(Icons.clear_rounded, color: Colors.white, size: 20),
          textFieldDecoration: BoxDecoration(
            borderRadius: BorderRadius.circular(16),
            color: Colors.indigo.shade200,
          ),
          inputDecoration: const InputDecoration(
            border: InputBorder.none,
            hintText: 'Search for your favoutie Anime',
            hintStyle: TextStyle(
              fontSize: 15,
              color: Colors.white,
            ),
            icon: Icon(
              Icons.search,
              color: Colors.white,
              size: 20,
            ),
          ),
          textFieldStyle: const TextStyle(
            fontSize: 15,
            color: Colors.white,
          ),
        ),
      ];

  List<Widget> example5() => [
        const Text('Right to left Direction'),
        DebounceTextfield(
          action: (enteredText) {
            print('Executed after 500 milliseconds from last change.');
            print(enteredText);
          },
          direction: TextDirection.rtl,
          inputDecoration: const InputDecoration.collapsed(
            hintText: 'أدخل النص هنا',
            hintStyle: TextStyle(
              fontSize: 13,
              color: Colors.black54,
            ),
          ),
          textFieldDecoration: BoxDecoration(
            borderRadius: BorderRadius.circular(16),
            color: Colors.black12,
          ),
        ),
      ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Debounce Textfield Demo'),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            children: <Widget>[
              ...example1(),
              ...example2(),
              ...example3(),
              ...example4(),
              ...example5()
            ],
          ),
        ),
      ),
    );
  }
}
